### PR TITLE
log-backup: restore meta kv with batch method

### PR DESCRIPTION
**Pull request by joccau**
Signed-off-by: joccau <zak.zhao@pingcap.com>

### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/37111

Problem Summary:
- Failed to restore index after PiTR to down-stream cluster.
- The current reason for this issue is that the same meta KV appears multiple times and is overwritten by the old KV record after restoring to the latest tableInfo.

### What is changed and how it works?
Set a lots of batch kv files and sort and restore these kv entries in batch files.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No code

### Manual test
- reproduce the test in https://github.com/pingcap/tidb/issues/37111
- The result after PiTR with old code
`
MySQL [(none)]> show create table test.sbtest43;
| sbtest43 | CREATE TABLE `sbtest43` (
  `id` int(11) NOT NULL AUTO_INCREMENT,
  `k` int(11) NOT NULL DEFAULT '0',
  `c` char(120) NOT NULL DEFAULT '',
  `pad` char(60) NOT NULL DEFAULT '',
  PRIMARY KEY (`id`) /*T![clustered_index] CLUSTERED */
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin AUTO_INCREMENT=1019915 |
`
- The result after PiTR with modified code
`
MySQL [(none)]> show create table test.sbtest43;
| sbtest43 | CREATE TABLE `sbtest43` (
  `id` int(11) NOT NULL AUTO_INCREMENT,
  `k` int(11) NOT NULL DEFAULT '0',
  `c` char(120) NOT NULL DEFAULT '',
  `pad` char(60) NOT NULL DEFAULT '',
  PRIMARY KEY (`id`) /*T![clustered_index] CLUSTERED */,
  KEY `k_43` (`k`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin AUTO_INCREMENT=1019915 |
`

The Key `K-43`(``) has been restored in above test.

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Review by Leavrth — APPROVED** [review]

File: br/pkg/restore/client.go [hunk 1]
```diff
@@ -2101,20 +2218,41 @@ func (rc *Client) RestoreMetaKVFile(
...
+) (uint64, uint64, error) {
+	var (
+		kvCount uint64
+		size    uint64
+	)
+
+	rc.rawKVClient.SetColumnFamily(columnFamily)
+
+	for _, entry := range entries {
+		log.Debug("txn entry", zap.Uint64("key-ts", entry.ts), zap.Int("txnKey-len", len(entry.e.Key)),
```
Comment (3pointer) [inline br/pkg/restore/client.go:2241]: maybe refine this debug log to give a clear message.

File: br/pkg/restore/client.go [hunk 2]
```diff
@@ -2101,20 +2218,41 @@ func (rc *Client) RestoreMetaKVFile(
...
+			zap.Int("txnValue-len", len(entry.e.Value)), zap.ByteString("txnKey", entry.e.Key))
+
+		newEntry, err := sr.RewriteKvEntry(&entry.e, columnFamily)
 		if err != nil {
-			log.Error("rewrite txn entry failed", zap.Int("klen", len(txnEntry.Key)),
-				logutil.Key("txn-key", txnEntry.Key))
+			log.Error("rewrite txn entry failed", zap.Int("klen", len(entry.e.Key)),
```
Comment (3pointer) [inline br/pkg/restore/client.go:2246]: ditto

File: br/pkg/restore/client.go [hunk 3]
```diff
@@ -2101,20 +2218,41 @@ func (rc *Client) RestoreMetaKVFile(
...
+				logutil.Key("txn-key", entry.e.Key))
 			return 0, 0, errors.Trace(err)
 		} else if newEntry == nil {
 			continue
 		}
 		log.Debug("rewrite txn entry", zap.Int("newKey-len", len(newEntry.Key)),
-			zap.Int("newValue-len", len(txnEntry.Value)), zap.ByteString("newkey", newEntry.Key))
+			zap.Int("newValue-len", len(entry.e.Value)), zap.ByteString("newkey", newEntry.Key))
```
Comment (3pointer) [inline br/pkg/restore/client.go:2253]: ditto

File: br/pkg/restore/client.go [hunk 4]
```diff
@@ -2047,41 +2077,128 @@ func (rc *Client) RestoreMetaKVFiles(
...
+				if err != nil {
+					return errors.Trace(err)
+				}
+				idx = i
+				rangeMin = f.MinTs
+				rangeMax = f.MaxTs
+			}
+		}
+
+		if i == len(files)-1 {
```
Comment (3pointer) [inline br/pkg/restore/client.go:2119]: should it be `if idx < len(files-1) {` 🤔

Comment (joccau) [reply br/pkg/restore/client.go:2119]: It represents that the loop reach the last file, we need to restoreBatch() the rest file at the end.

**Review by 3pointer — APPROVED** [review]
LGTM

**Pull request merged**